Code/GraphMol/AddHs.cpp

```diff
@@ -855,6 +909,38 @@ bool shouldRemoveH(const RWMol &mol, const Atom *atom,
   return removeIt;
 }
 
+// Do not remove H atoms that are part of SGroups that only contain H atoms.
+void filter_sgroup_emptying_hydrogens(const ROMol &mol,
+                                      boost::dynamic_bitset<> &atomsToRemove) {
+  for (const auto &sg : getSubstanceGroups(mol)) {
+    const auto &atoms = sg.getAtoms();
+    const auto &patoms = sg.getParentAtoms();
+
+    // If the SGroup already didn't have atoms, we don't care about it
+    if (atoms.empty() && patoms.empty()) {
+      continue;
+    }
+
+    auto would_remove_atom = [&atomsToRemove](const auto idx) {
+      return atomsToRemove[idx];
+    };
+
+    auto no_atoms = atoms.empty() ||
+                    std::all_of(atoms.begin(), atoms.end(), would_remove_atom);
+    auto no_patoms = patoms.empty() || std::all_of(patoms.begin(), patoms.end(),
+                                                   would_remove_atom);
+
+    if (no_atoms && no_patoms) {
+      for (auto atom : atoms) {
+        atomsToRemove.set(atom, false);
+      }
+      for (auto patom : patoms) {
+        atomsToRemove.set(patom, false);
+      }
+    }
```

Comment:
Minor, this form would do a bit less work under some circumstances:
```suggestion
    if (no_atoms) {
      auto no_patoms = patoms.empty() || std::all_of(patoms.begin(), patoms.end(),
                                                     would_remove_atom);
      if (no_patoms) {
        for (auto atom : atoms) {
          atomsToRemove.set(atom, false);
        }
        for (auto patom : patoms) {
          atomsToRemove.set(patom, false);
        }
      }
    }
```

Reply:
Yeah, makes sense.